src/number_theory/modular.lean

```diff
@@ -0,0 +1,377 @@
+  tendsto (λ g : {g : SL(2, ℤ) // g 1 = p}, |((g : SL(2, ℤ)) • z).re|)
+    cofinite at_top :=
+begin
+  suffices : tendsto (λ g : (λ g : SL(2, ℤ), g 1) ⁻¹' {p}, (((g : SL(2, ℤ)) • z).re))
+    cofinite (cocompact ℝ),
+  { exact tendsto_norm_cocompact_at_top.comp this },
+  have : ((p 0 : ℝ) ^ 2 + p 1 ^ 2)⁻¹ ≠ 0,
+  { apply inv_ne_zero,
+    exact_mod_cast hp.sq_add_sq_ne_zero },
+  let f := homeomorph.mul_right₀ _ this,
+  let ff := homeomorph.add_right (((p 1:ℂ)* z - p 0) / ((p 0 ^ 2 + p 1 ^ 2) * (p 0 * z + p 1))).re,
+  convert ((f.trans ff).closed_embedding.tendsto_cocompact).comp (tendsto_lc_row0 hp),
+  ext g,
+  change ((g : SL(2, ℤ)) • z).re = (lc_row0 p ↑(↑g : SL(2, ℝ))) / (p 0 ^ 2 + p 1 ^ 2)
+  + (((p 1:ℂ )* z - p 0) / ((p 0 ^ 2 + p 1 ^ 2) * (p 0 * z + p 1))).re,
+  exact_mod_cast (congr_arg complex.re (smul_eq_lc_row0_add hp z g.2))
+end
+
+end tendsto_lemmas
+
+section fundamental_domain
+
+local attribute [simp] coe_smul re_smul
+
+/-- For `z : ℍ`, there is a `g : SL(2,ℤ)` maximizing `(g•z).im` -/
+lemma exists_g_with_max_im (z : ℍ) :
+  ∃ g : SL(2, ℤ), ∀ g' : SL(2, ℤ), (g' • z).im ≤ (g • z).im :=
+begin
+  classical,
+  let s : set (fin 2 → ℤ) := {cd | is_coprime (cd 0) (cd 1)},
+  have hs : s.nonempty := ⟨![1, 1], is_coprime_one_left⟩,
+  obtain ⟨p, hp_coprime, hp⟩ :=
+    filter.tendsto.exists_within_forall_le hs (tendsto_norm_sq_coprime_pair z),
+  obtain ⟨g, -, hg⟩ := bottom_row_surj hp_coprime,
+  refine ⟨g, λ g', _⟩,
+  rw [im_smul_eq_div_norm_sq, im_smul_eq_div_norm_sq, div_le_div_left],
+  { simpa [← hg] using hp (g' 1) (bottom_row_coprime g') },
+  { exact z.im_pos },
+  { exact norm_sq_denom_pos g' z },
+  { exact norm_sq_denom_pos g z },
+end
+
+/-- Given `z : ℍ` and a bottom row `(c,d)`, among the `g : SL(2,ℤ)` with this bottom row, minimize
+  `|(g•z).re|`.  -/
+lemma exists_g_with_given_cd_and_min_re (z:ℍ) {cd : fin 2 → ℤ} (hcd : is_coprime (cd 0) (cd 1)) :
+  ∃ g : SL(2,ℤ), ↑ₘg 1 = cd ∧ (∀ g' : SL(2,ℤ), ↑ₘg 1 = ↑ₘg' 1 →
+  |(g • z).re| ≤ |(g' • z).re|) :=
+begin
+  haveI : nonempty {g : SL(2, ℤ) // g 1 = cd} := let ⟨x, hx⟩ := bottom_row_surj hcd in ⟨⟨x, hx.2⟩⟩,
+  obtain ⟨g, hg⟩ := filter.tendsto.exists_forall_le (tendsto_abs_re_smul z hcd),
+  refine ⟨g, g.2, _⟩,
+  { intros g1 hg1,
+    have : g1 ∈ ((λ g : SL(2, ℤ), g 1) ⁻¹' {cd}),
+    { rw [set.mem_preimage, set.mem_singleton_iff],
+      exact eq.trans hg1.symm (set.mem_singleton_iff.mp (set.mem_preimage.mp g.2)) },
+    exact hg ⟨g1, this⟩ },
+end
+
+/-- The matrix `T = [[1,1],[0,1]]` as an element of `SL(2,ℤ)` -/
+def T : SL(2,ℤ) := ⟨![![1, 1], ![0, 1]], by norm_num [matrix.det_fin_two]⟩
+
+/-- The matrix `T' (= T⁻¹) = [[1,-1],[0,1]]` as an element of `SL(2,ℤ)` -/
+def T' : SL(2,ℤ) := ⟨![![1, -1], ![0, 1]], by norm_num [matrix.det_fin_two]⟩
+
+/-- The matrix `S = [[0,-1],[1,0]]` as an element of `SL(2,ℤ)` -/
+def S : SL(2,ℤ) := ⟨![![0, -1], ![1, 0]], by norm_num [matrix.det_fin_two]⟩
```

Comment:
It would be nice to have some simp-lemmas about these matrices. But maybe let's save that for a future PR. (At least knowing things like `S  * S = -1` seems worth having as a lemma.)